### PR TITLE
Hammer

### PR DESCRIPTION
**Pull request by rahul1aggarwal**

### CODE REVIEW

File: src/rgw/rgw_common.h [hunk 1]
```diff
@@ -993,78 +993,160 @@ inline ostream& operator<<(ostream& out, const rgw_obj_key &o) {
...
+    inline void set_acl_copy_override(bool val)
+    {
+      _acl_copy_override = val;
+    }
+
+
+    authorization_method(bool method, bool action, bool url_token) :
+      _token_validation(method),
+      _copy_action(action),
+      _url_type_token(url_token) { }
```
Comment (shivanshu21) [inline src/rgw/rgw_common.h]: Minor: I forgot to add acl_override bools here for initialization. Maybe we should put them in constructor.

File: src/rgw/rgw_main.cc [hunk 2]
```diff
@@ -558,6 +558,41 @@ static int process_request(RGWRados *store, RGWREST *rest, RGWRequest *req, RGWC
...
+  const char* jcs_metadata_directive = s->info.env->get("HTTP_X_JCS_METADATA_DIRECTIVE");
+  const char* amz_copy_source = s->info.env->get("HTTP_X_AMZ_COPY_SOURCE");
+  const char* jcs_copy_source = s->info.env->get("HTTP_X_JCS_COPY_SOURCE");
+
+  if(
+      (
+        (amz_metadata_directive != NULL && !strcmp(amz_metadata_directive, "COPY")) 
+        || (jcs_metadata_directive != NULL && !strcmp(jcs_metadata_directive, "COPY"))
+      ) 
+      ||
```
Comment (shivanshu21) [inline src/rgw/rgw_main.cc]: Make it AND. In a rare case it it possible that x-amz-copy-source is not set but metadata directive is set to copy. RGW will not take this case as copy but we will attempt validation for copy case during authorize()

File: src/rgw/rgw_main.cc [hunk 3]
```diff
@@ -558,6 +558,41 @@ static int process_request(RGWRados *store, RGWREST *rest, RGWRequest *req, RGWC
...
+
+  if(
+      (
+        (amz_metadata_directive != NULL && !strcmp(amz_metadata_directive, "COPY")) 
+        || (jcs_metadata_directive != NULL && !strcmp(jcs_metadata_directive, "COPY"))
+      ) 
+      || 
+      (amz_copy_source != NULL || jcs_copy_source != NULL)
+    ) {
+    (s->auth_method).set_copy_action(true);
```
Comment (shivanshu21) [inline src/rgw/rgw_main.cc:586]: Make the condition AND as mentioned in the previous comment. Lets only set copy_action to true when copy source is provided.

File: src/rgw/rgw_main.cc [hunk 4]
```diff
@@ -747,13 +783,15 @@ static int civetweb_callback(struct mg_connection *conn) {
           dout(1) << "DSS INFO: CIVETWEB HEADER NAME: " << name_str << dendl;
```
Comment (shivanshu21) [inline src/rgw/rgw_main.cc:783]: Minor: Do we still want to print all the headers?

File: src/rgw/rgw_common.h (hunk 1 above)
Comment (rahul1aggarwal) [inline src/rgw/rgw_common.h]: ok. will do

File: src/rgw/rgw_main.cc (hunk 2 above)
Comment (rahul1aggarwal) [inline src/rgw/rgw_main.cc]: ok

File: src/rgw/rgw_main.cc (hunk 4 above)
Comment (rahul1aggarwal) [inline src/rgw/rgw_main.cc:783]: yes, let's print the headers always. We will find a nicer way when we revisit logging in the coming days.

**Pull request merged**